From 380aae25ffb991fb4c545b2f87868154e6a5c9fe Mon Sep 17 00:00:00 2001
From: Vignesh Viswanathan <quic_viswanat@quicinc.com>
Date: Wed, 23 Oct 2024 17:58:26 +0530
Subject: [PATCH] phy: qcom-qusb2: Set device driver_data early in probe

In qusb2_phy_probe, qphy is set as driver_data to the device towards the
end of the probe. pm_runtime ops registered for this device is enabled
before the driver_data is set but in the pm_ops, dev_get_drvdata is used
to get qphy and dereferenced without NULL checks. This causes a NULL
pointer dereference if pm_ops are invoked before the probe is completed.

Move setting driver_data to earlier in the probe as soon as qphy is
allocated.

Change-Id: I848521140c4524196ff4949c8a33d93c000d33c8
Signed-off-by: Vignesh Viswanathan <quic_viswanat@quicinc.com>
---
 drivers/phy/qualcomm/phy-qcom-qusb2.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/phy/qualcomm/phy-qcom-qusb2.c b/drivers/phy/qualcomm/phy-qcom-qusb2.c
index aee9e1ceefbd..d8f79316b621 100644
--- a/drivers/phy/qualcomm/phy-qcom-qusb2.c
+++ b/drivers/phy/qualcomm/phy-qcom-qusb2.c
@@ -997,6 +997,8 @@ static int qusb2_phy_probe(struct platform_device *pdev)
 	qphy = devm_kzalloc(dev, sizeof(*qphy), GFP_KERNEL);
 	if (!qphy)
 		return -ENOMEM;
+
+	dev_set_drvdata(dev, qphy);
 	or = &qphy->overrides;
 
 	qphy->base = devm_platform_ioremap_resource(pdev, 0);
@@ -1109,7 +1111,6 @@ static int qusb2_phy_probe(struct platform_device *pdev)
 	}
 	qphy->phy = generic_phy;
 
-	dev_set_drvdata(dev, qphy);
 	phy_set_drvdata(generic_phy, qphy);
 
 	phy_provider = devm_of_phy_provider_register(dev, of_phy_simple_xlate);
-- 
2.34.1

